### PR TITLE
Lock codegen to avoid crashes on when running multithreaded code

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

**Comment by maleadt** [comment]
The unlock should be in the `finally` of a `try/catch`; in case the GPUCompiler errors, we still need to unlock or we'd risk breaking Julia.

**Comment by maleadt** [comment]
Segfault looks related?

File: src/driver.jl [hunk 1]
```diff
@@ -85,177 +85,183 @@ function codegen(output::Symbol, @nospecialize(job::CompilerJob);
 
 
     ## LLVM IR
+    # Lock codegen to prevent races on the context
+    # the lock is released in a finally block
+    ccall(:jl_typeinf_end, Cvoid, ())
```
Comment (maleadt) [inline src/driver.jl:90]: You only call _end.

Comment (vchuravy) [reply src/driver.jl:90]: .............

**Comment by maleadt** [comment]
https://github.com/JuliaLang/julia/pull/38732 broke LLVM.jl, https://github.com/maleadt/LLVM.jl/pull/219, so this may be the cause for the errors here. It resulted in illegal memory accesses in CUDA.jl, which is pretty wild.

**Review by maleadt — APPROVED** [review]
LGTM.

**Comment by vchuravy** [comment]
Diff is much nicer to check with whitespace changes off https://github.com/JuliaGPU/GPUCompiler.jl/pull/120/files?diff=unified&w=1

**Comment by maleadt** [comment]
```
From worker 5:	signal (11): Segmentation fault
From worker 5:	in expression starting at none:1
From worker 5:	jl_lock_frame_push at /buildworker/worker/package_linux64/build/src/locks.h:65 [inlined]
From worker 5:	jl_mutex_lock at /buildworker/worker/package_linux64/build/src/locks.h:89 [inlined]
From worker 5:	inst_datatype_inner at /buildworker/worker/package_linux64/build/src/jltypes.c:1253
From worker 5:	jl_inst_arg_tuple_type at /buildworker/worker/package_linux64/build/src/jltypes.c:1424
From worker 5:	arg_type_tuple at /buildworker/worker/package_linux64/build/src/gf.c:1823 [inlined]
From worker 5:	jl_lookup_generic_ at /buildworker/worker/package_linux64/build/src/gf.c:2350 [inlined]
From worker 5:	jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2402
From worker 5:	#check_cache#100 at /var/lib/buildkite-agent/builds/9700k-gpuci3-julia-csail-mit-edu/julialang/gpucompiler-dot-jl/src/cache.jl:48
From worker 5:	check_cache##kw at /var/lib/buildkite-agent/builds/9700k-gpuci3-julia-csail-mit-edu/julialang/gpucompiler-dot-jl/src/cache.jl:15
From worker 5:	unknown function (ip: 0x7f5c366caa69)
From worker 5:	#cached_compilation#101 at /root/.julia/packages/GPUArrays/WV76E/src/host/math.jl:5 [inlined]
From worker 5:	#cached_compilation#101 at /var/lib/buildkite-agent/builds/9700k-gpuci3-julia-csail-mit-edu/julialang/gpucompiler-dot-jl/src/cache.jl:0
From worker 5:	unknown function (ip: 0x7f5c366c4d86)
```

This looks related...

**Comment by maleadt** [comment]
Superseded by https://github.com/JuliaGPU/GPUCompiler.jl/pull/120.

**Pull request closed without merging**